packages/plugins/inline-trace/src/index.ts

```diff
@@ -0,0 +1,313 @@
+import { isAsyncIterable, Plugin, YogaInitialContext } from 'graphql-yoga'
+import { GraphQLError, ResponsePath } from 'graphql'
+import { google, Trace } from 'apollo-reporting-protobuf'
+
+interface InlineTraceContext {
+  startHrTime: [number, number]
+  rootNode: Trace.Node
+  trace: Trace
+  nodes: Map<string, Trace.Node>
+  /**
+   * graphql-js can continue to execute more fields indefinitely after
+   * `execute()` resolves. That's because parallelism on a selection set
+   * is implemented using `Promise.all`, and as soon as one field
+   * throws an error, the combined Promise resolves, but there's no
+   * "cancellation" of the rest of Promises/fields in `Promise.all`.
+   */
+  stopped: boolean
+}
+
+export interface InlineTracePluginOptions {
+  /**
+   * Format errors before being sent for tracing. Beware that only the error
+   * `message` and `extensions` can be changed.
+   *
+   * Return `null` to skip reporting error.
+   */
+  rewriteError?: (err: GraphQLError) => GraphQLError | null
+}
+
+/**
+ * Produces Apollo's base64 trace protocol containing timing, resolution and
+ * errors information.
+ *
+ * The output is placed in `extensions.ftv1` of the GraphQL result.
+ *
+ * The Apollo Gateway utilizes this data to construct the full trace and submit
+ * it to Apollo's usage reporting ingress.
+ */
+export function useInlineTrace(
+  options: InlineTracePluginOptions = {},
+): Plugin<YogaInitialContext> {
+  const ctxForReq = new WeakMap<Request, InlineTraceContext>()
+
+  return {
+    onRequest({ request }) {
```

Comment:
Maybe we can let user decide to set the "request" object for WeakMap then make this a server agnostic Envelop plugin. Then we can create a Yoga variation of it. 
What do you think?

Reply:
The only thing the request has to have is the `apollo-federation-include-trace` header. Maybe move it envelop and have some function like: `function shouldTrace(ctx): boolean` that dictates whether the tracing should occur.

Reply:
That's exactly what I had on mind at first!

Reply:
See https://github.com/dotansimha/graphql-yoga/pull/1621#issuecomment-1220464450.